core/runtime/src/main/java/org/apache/isis/core/runtime/action/invocation/ActionInvocationFacetWithQueryResultsCache.java

```diff
@@ -0,0 +1,87 @@
+package org.apache.isis.core.runtime.action.invocation;
+
+import java.util.concurrent.Callable;
+
+import org.apache.isis.applib.services.queryresultscache.QueryResultsCache;
+import org.apache.isis.core.metamodel.adapter.ObjectAdapter;
+import org.apache.isis.core.metamodel.facetapi.DecoratingFacet;
+import org.apache.isis.core.metamodel.facets.actions.action.invocation.ActionInvocationFacet;
+import org.apache.isis.core.metamodel.facets.actions.action.invocation.ActionInvocationFacetAbstract;
+import org.apache.isis.core.metamodel.spec.ObjectSpecification;
+import org.apache.isis.core.metamodel.spec.feature.ObjectAction;
+import org.apache.isis.core.runtime.system.context.IsisContext;
+
+public class ActionInvocationFacetWithQueryResultsCache extends ActionInvocationFacetAbstract implements DecoratingFacet<ActionInvocationFacet> {
+
+    final ActionInvocationFacet underlyingFacet;
+
+    @Override
+    public ActionInvocationFacet getDecoratedFacet() {
+        return this.underlyingFacet;
+    }
+
+    public ActionInvocationFacetWithQueryResultsCache(
+            final ActionInvocationFacet underlyingFacet) {
+        super(underlyingFacet.getFacetHolder());
+        this.underlyingFacet = underlyingFacet;
+    }
+
+    @Override
+    public ObjectAdapter invoke(
+            final ObjectAction owningAction,
+            final ObjectAdapter targetAdapter,
+            final ObjectAdapter[] argumentAdapters) {
+
+        final ObjectAdapter result = getQueryResultsCache().execute(new Callable<ObjectAdapter>() {
+
+            @Override
+            public ObjectAdapter call() throws Exception {
+                return ActionInvocationFacetWithQueryResultsCache.this.underlyingFacet.invoke(owningAction, targetAdapter, argumentAdapters);
+            }
+        }, ActionInvocationFacetWithQueryResultsCache.class, "invoke", owningAction, targetAdapter, targetAdapter);
```

Comment:
There could be a rare issue with this... the arguments are used to create a cache key, and because you've used the AIFWQRC.class and "invoke" itself, rather than that of the target method that's ultimately called, it's conceivable that caching could occur incorrectly.

eg:
someObject.foo(int, int) -> someObject.bar(int, int)

If both "foo" and "bar" were marked to be cached, then the "foo" cache could be used for "bar".